js/data/bucket.js

```diff
@@ -132,26 +127,32 @@ Bucket.prototype.populateBuffers = function() {
  * @param {number} vertexLength The number of vertices that will be inserted to the buffer.
  */
 Bucket.prototype.makeRoomFor = function(programName, numVertices) {
-    var groups = this.elementGroups[programName];
+    var groups = this.arrays[programName];
     var currentGroup = groups.length && groups[groups.length - 1];
 
-    if (!currentGroup || currentGroup.vertexLength + numVertices > 65535) {
-        var layoutArrays = this.arrays[programName].layout;
-        var vertexArray = layoutArrays.vertex;
-        var elementArray = layoutArrays.element;
-        var secondElementArray = layoutArrays.secondElement;
+    if (!currentGroup || currentGroup.layout.vertex.length + numVertices > 65535) {
+
+        var arrayTypes = this.arrayTypes[programName];
+        var VertexArrayType = arrayTypes.layout.vertex;
+        var ElementArrayType = arrayTypes.layout.element;
+        var SecondElementArrayType = arrayTypes.layout.secondElement;
 
         currentGroup = {
-            vertexStartIndex: vertexArray.length,
-            elementStartIndex: elementArray && elementArray.length,
-            secondElementStartIndex: secondElementArray && secondElementArray.length,
-            elementLength: 0,
-            vertexLength: 0,
-            secondElementLength: 0,
-            elementOffset: elementArray && elementArray.length * elementArray.bytesPerElement,
-            secondElementOffset: secondElementArray && secondElementArray.length * secondElementArray.bytesPerElement,
-            vertexOffset: vertexArray && vertexArray.length * vertexArray.bytesPerElement
+            index: groups.length,
+            layout: {},
+            paint: {}
         };
+
+        currentGroup.layout.vertex = new VertexArrayType();
+        if (ElementArrayType) currentGroup.layout.element = new ElementArrayType();
+        if (SecondElementArrayType) currentGroup.layout.secondElement = new SecondElementArrayType();
+
+        for (var i = 0; i < this.childLayers.length; i++) {
+            var layerName = this.childLayers[i].id;
+            var PaintVertexArrayType = arrayTypes.paint[layerName];
+            currentGroup.paint[layerName] = new PaintVertexArrayType();
+        }
+
         groups.push(currentGroup);
```

Comment:
The terminology in here is a little confusing. We've "eliminated element groups" but still have an entity that we're referring to as a "group" but putting into a property called "array." Is there an easy way to shore up the terminology here?